Imlement Stops report

Abyss777 opened the pull request:
About stops:
- "Stop" is a gap between trips
- It's minimal duration can't be less than `report.trip.minimalParkingDuration` except one case: it is in the end of period and after a trip
- Stop address and coordinates are taken from the first position
- Also it has following fields: `duration`, `engineHours`, `spentFuel`

![image](https://user-images.githubusercontent.com/5688080/27119594-e7cb0618-50f9-11e7-8f6a-3e05896b2554.png)

About code:
- Moved trips detector to `ReportUtils` and adapted it to generate stops
- Moved `spentFuel` to `BaseReport`
- Implemented `TripsConfig` object just because of limit in function parameters count
- Extended tests for some edge cases

More comments in the code

Abyss777 commented on src/org/traccar/reports/ReportUtils.java:
```diff
@@ -123,4 +127,186 @@ public static void processTemplateWithSheets(InputStream templateStream, OutputS
+                                            .getFixTime().getTime() - positions.get(previousStartParkingIndex)
+                                            .getFixTime().getTime();
+                                    if (previousParkingDuration >= tripsConfig.getMinimalParkingDuration()) {
+                                        result.add(calculateStop(positions, previousStartParkingIndex,
```
Add Stop before first trip if it is long enough

Abyss777 commented on src/org/traccar/reports/ReportUtils.java:
```diff
@@ -123,4 +127,186 @@ public static void processTemplateWithSheets(InputStream templateStream, OutputS
+                    startParkingIndex = -1;
+                }
+            }
+            if (result.isEmpty() && !trips) {
```
If we did not detected any trips then whole(or piece) period might be a stop

tananaev commented on test/org/traccar/reports/TripsAndStopsTest.java:
```diff
@@ -0,0 +1,199 @@
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class TripsAndStopsTest extends BaseTest {
```
Name of the test file should exactly match the class that you are testing (and have `Test` suffix).

tananaev commented on src/org/traccar/reports/ReportUtils.java:
```diff
@@ -123,4 +127,186 @@ public static void processTemplateWithSheets(InputStream templateStream, OutputS
+    }
+
+    public static Collection<BaseReport> detectTripsAndStops(TripsConfig tripsConfig,
+             boolean ignoreOdometer, double speedThreshold,
```
There is indentation problem somewhere here.

Abyss777 commented:
Fixed indentation and moved tests to correct place `ReportUtilsTest`

Agree, it is unreadable and agree do not think there is a way to make it clear. I can try add comments to algorithm, but I think make it looks worse.

tananaev commented:
Merged, thanks. Is web part coming as well?

kirantpatil commented:
Does it also shows stop points on the map ?

Abyss777 commented:
@kirantpatil The question is more related to this PR https://github.com/tananaev/traccar-web/pull/504
But yes, stop points are shown on the map.

kirantpatil commented:
Great and Thanks @Abyss777
